### PR TITLE
Adding Seq for opcode and param translation in TLMessages

### PR DESCRIPTION
**Pull request by srivatsa611y**
We are adding methods to decode the current Opcode and Param information on the tilelink bundle.

The way it is done might be messy and hoping for the reviewers for better ways.

### CODE REVIEW

File: src/main/scala/tilelink/Bundles.scala [hunk 1]
```diff
@@ -134,8 +137,59 @@ final class TLBundleA(params: TLBundleParameters)
...
+      is(0.U) {printf("OP: PutFullData      ")} 
+      is(1.U) {printf("OP: PutPartialData   ")}
+      is(2.U) {printf("OP: ArithmeticData   ")} 
+      is(3.U) {printf("OP: LogicalData      ")} 
+      is(4.U) {printf("OP: Get              ")}
+      is(5.U) {printf("OP: Hint             ")}
+      is(6.U) {printf("OP: AcquireBlock     ")}
+      is(7.U) {printf("OP: AcquirePerm      ")}
+    }
+  }
```
Comment (seldridge) [inline src/main/scala/tilelink/Bundles.scala]: A functional style can help with the clutter. One equivalent approach would be:
```scala
  def opcodePrint(): Unit =
    Seq("PutFullData", "PutPartialData", "ArithmeticData", "LogicalData", "Get", "Hint", "AcquireBlock", "AcquirePerm")
      .zipWithIndex
      .map{ case (str, idx) => when (opcode === idx.U) { printf(s"OP: $str\n") } }
```

Comment (srivatsa611y) [reply src/main/scala/tilelink/Bundles.scala]: Can this be done in runtime, because thats where we will be using it.

Comment (seldridge) [reply src/main/scala/tilelink/Bundles.scala]: Both methods will statically emit circuits with `printf`s that trip on the conditions specified. So, they will generate hardware `printf`s that are usable at runtime.

Note: there is a subtle difference here in that `switch`/`is` produces a `when`/`else` nesting while the explicit `when` blocks of the functional style are all flat. Functionally, there's no difference, though.

The original is doing something like:
```
    node _T_33 = eq(UInt<1>("h00"), UInt<1>("h01")) 
    when _T_33 : 
      node _T_34 = bits(reset, 0, 0) 
      node _T_35 = eq(_T_34, UInt<1>("h00")) 
      when _T_35 : 
        printf(clock, UInt<1>(1), "OP: PutFullData      ") 
    else : 
      node _T_36 = eq(UInt<1>("h01"), UInt<1>("h01")) 
      when _T_36 : 
        node _T_37 = bits(reset, 0, 0) 
        node _T_38 = eq(_T_37, UInt<1>("h00")) 
        when _T_38 : 
          printf(clock, UInt<1>(1), "OP: PutPartialData   ") 
```

The `map` is generating:
```
    node _T_9 = eq(UInt<1>("h01"), UInt<1>("h00")) 
    when _T_9 : 
      node _T_10 = bits(reset, 0, 0) 
      node _T_11 = eq(_T_10, UInt<1>("h00")) 
      when _T_11 : 
        printf(clock, UInt<1>(1), "OP: PutFullData\n") 
    node _T_12 = eq(UInt<1>("h01"), UInt<1>("h01")) 
    when _T_12 : 
      node _T_13 = bits(reset, 0, 0) 
      node _T_14 = eq(_T_13, UInt<1>("h00")) 
      when _T_14 : 
        printf(clock, UInt<1>(1), "OP: PutPartialData\n") 
```

**Comment by hcook** [comment]
Why are there two distinct print functions instead of one a single one that prints something about every field in the channel?

**Comment by hcook** [comment]
You should use the defs in the `TLMessages` object, rather than hardcoding the values in two places: https://github.com/freechipsproject/rocket-chip/blob/master/src/main/scala/tilelink/Bundles.scala#L16

We might change the encoding in the future!

**Comment by srivatsa611y** [comment]
yea, we wanted your reviews. the print is all being moved out of rocketchip. We are planning to have maps as suggested by @seldridge

**Comment by srivatsa611y** [comment]
Changes done as requested.

**Comment by srivatsa611y** [comment]
@hcook  @terpstra  Can you guys please review the pr. I have done the changes as requested.

File: src/main/scala/tilelink/Bundles.scala [hunk 2]
```diff
@@ -44,6 +44,41 @@ object TLMessages
...
   def bcResponse = Vec(AccessAck, AccessAck, AccessAckData, AccessAckData, AccessAckData, HintAck, ProbeAck, ProbeAck)
+  
+  def a = Seq( ("PutFullData",TLPermissions.PermMsgReserved),
+               ("PutPartialData",TLPermissions.PermMsgReserved),
+               ("ArithmeticData",TLAtomics.ArithMsg),
+               ("LogicalData",TLAtomics.LogicMsg),
+               ("Get",TLPermissions.PermMsgReserved),
+               ("Hint",TLHints.HintsMsg),
+               ("AcquireBlock",TLPermissions.PermMsgGrow),
+               ("AcquirePerm",TLPermissions.PermMsgReserved))
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: Oh. This should be PermMsgGrow.

File: src/main/scala/tilelink/Bundles.scala [hunk 3]
```diff
@@ -44,6 +44,41 @@ object TLMessages
...
+
+  def b = Seq( ("PutFullData",TLPermissions.PermMsgReserved),
+               ("PutPartialData",TLPermissions.PermMsgReserved),
+               ("ArithmeticData",TLPermissions.PermMsgReserved),
+               ("LogicalData",TLPermissions.PermMsgReserved),
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: This should be LogicMsg

File: src/main/scala/tilelink/Bundles.scala [hunk 4]
```diff
@@ -44,6 +44,41 @@ object TLMessages
...
+               ("ArithmeticData",TLAtomics.ArithMsg),
+               ("LogicalData",TLAtomics.LogicMsg),
+               ("Get",TLPermissions.PermMsgReserved),
+               ("Hint",TLHints.HintsMsg),
+               ("AcquireBlock",TLPermissions.PermMsgGrow),
+               ("AcquirePerm",TLPermissions.PermMsgReserved))
+
+  def b = Seq( ("PutFullData",TLPermissions.PermMsgReserved),
+               ("PutPartialData",TLPermissions.PermMsgReserved),
+               ("ArithmeticData",TLPermissions.PermMsgReserved),
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: This should be ArithMsg

File: src/main/scala/tilelink/Bundles.scala [hunk 5]
```diff
@@ -44,6 +44,41 @@ object TLMessages
...
+               ("LogicalData",TLPermissions.PermMsgReserved),
+               ("Get",TLPermissions.PermMsgReserved),
+               ("Hint",TLPermissions.PermMsgCap),
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: Should be HintsMsg

File: src/main/scala/tilelink/Bundles.scala [hunk 6]
```diff
@@ -44,6 +44,41 @@ object TLMessages
...
+               ("Probe",TLPermissions.PermMsgReserved))
```
Comment (terpstra) [inline src/main/scala/tilelink/Bundles.scala]: Should be PermsMsg. something. can't review now.

File: src/main/scala/tilelink/Bundles.scala (hunk 2 above)
Comment (srivatsa611y) [reply src/main/scala/tilelink/Bundles.scala]: Well they are for the reserved fields. So i put them all in 1 object and use the same seq in other places. Do you want me to change is accordingly?

Comment (srivatsa611y) [reply src/main/scala/tilelink/Bundles.scala]: OOh sorry my bad. I'll cross verify it

Comment (srivatsa611y) [reply src/main/scala/tilelink/Bundles.scala]: The AcquirePerm Is not there in the TileLink Spec. Draft 1.7

**Pull request merged**